### PR TITLE
asm/f32:  Adding dot product asm functions.

### PR DESCRIPTION
**Pull request by Kunde21**
Migrated from gonum/internal#58 with uintptr corrections.

### CODE REVIEW

File: internal/asm/f32/ddotinc_amd64.s [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+	MOVQ incX+56(FP), INC_X // INC_X = incX * sizeof(float32)
+	SHLQ $2, INC_X
+	MOVQ incY+64(FP), INC_Y // INC_Y = incY * sizeof(float32)
+	SHLQ $2, INC_Y
+
+	MOVQ LEN, TAIL
+	ANDQ $3, TAIL  // TAIL = LEN % 4
+	SHRQ $2, LEN   // LEN = floor( LEN / 4 )
+	JZ   dot_tail  // if LEN == 0 { goto dot_tail_start }
```
Comment (kortschak) [inline internal/asm/f32/ddotinc_amd64.s]: `goto dot_tail`

File: internal/asm/f32/ddotunitary_amd64.s [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	CMPQ    y_len+32(FP), LEN
+	CMOVQLE y_len+32(FP), LEN
+	PXOR    SUM, SUM             // psum = 0
+	CMPQ    LEN, $0
+	JE      dot_end
+
+	XORQ IDX, IDX
+	MOVQ Y_PTR, DX
+	ANDQ $0xF, DX    // Align on 16-byte boundary for ADDPS
+	JZ   dot_no_trim // if DX == 0 { goto axpy_no_trim }
```
Comment (kortschak) [inline internal/asm/f32/ddotunitary_amd64.s]: `goto dot_no_trim`

File: internal/asm/f32/dotunitary_amd64.s [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	MOVQ    x_len+8(FP), LEN     // LEN = min( len(x), len(y) )
+	CMPQ    y_len+32(FP), LEN
+	CMOVQLE y_len+32(FP), LEN
+	CMPQ    LEN, $0
+	JE      dot_end
+
+	XORQ IDX, IDX
+	MOVQ Y_PTR, DX
+	ANDQ $0xF, DX    // Align on 16-byte boundary for MULPS
+	JZ   dot_no_trim // if DX == 0 { goto axpy_no_trim }
```
Comment (kortschak) [inline internal/asm/f32/dotunitary_amd64.s]: `goto dot_no_trim`

File: internal/asm/f32/benchDot_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+func BenchmarkDotUnitary(t *testing.B) {
+	const name = "DotUnitary"
+	for _, v := range []int64{1, 2, 3, 4, 5, 10, 100, 1e3, 5e3, 1e4, 5e4} {
+		t.Run(fmt.Sprintf("%s-%d", name, v), func(b *testing.B) {
+			x, y := x[:v], y[:v]
+			b.SetBytes(32 * v)
+			for i := 0; i < b.N; i++ {
+				DotUnitary(x, y)
```
Comment (kortschak) [inline internal/asm/f32/benchDot_test.go]: This needs to assign to a global.

File: internal/asm/f32/benchDot_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func BenchmarkDdotUnitary(t *testing.B) {
+	const name = "DdotUnitary"
+	for _, v := range []int64{1, 2, 3, 4, 5, 10, 100, 1e3, 5e3, 1e4, 5e4} {
+		t.Run(fmt.Sprintf("%s-%d", name, v), func(b *testing.B) {
+			x, y := x[:v], y[:v]
+			b.SetBytes(32 * v)
+			for i := 0; i < b.N; i++ {
+				DdotUnitary(x, y)
```
Comment (kortschak) [inline internal/asm/f32/benchDot_test.go]: This too.

File: internal/asm/f32/benchDot_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	for _, tt := range incsDot {
+		for _, inc := range tt.inc {
+			t.Run(fmt.Sprintf("%s-%d-inc(%d)", name, tt.len, inc), func(b *testing.B) {
+				b.SetBytes(int64(32 * tt.len))
+				idx := 0
+				if inc < 0 {
+					idx = (-tt.len + 1) * inc
+				}
+				for i := 0; i < b.N; i++ {
+					DotInc(x, y, uintptr(tt.len), uintptr(inc), uintptr(inc), uintptr(idx), uintptr(idx))
```
Comment (kortschak) [inline internal/asm/f32/benchDot_test.go]: And here.

File: internal/asm/f32/benchDot_test.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	for _, tt := range incsDot {
+		for _, inc := range tt.inc {
+			t.Run(fmt.Sprintf("%s-%d-inc(%d)", name, tt.len, inc), func(b *testing.B) {
+				b.SetBytes(int64(32 * tt.len))
+				idx := 0
+				if inc < 0 {
+					idx = (-tt.len + 1) * inc
+				}
+				for i := 0; i < b.N; i++ {
+					DdotInc(x, y, uintptr(tt.len), uintptr(inc), uintptr(inc), uintptr(idx), uintptr(idx))
```
Comment (kortschak) [inline internal/asm/f32/benchDot_test.go]: And here.

**Comment by kortschak** [comment]
Are these asm PRs ready to be merged?

**Comment by Kunde21** [comment]
Yes, I was just giving @vladimir-ch an opportunity to take a look.

**Comment by vladimir-ch** [comment]
I will be able to take a look starting Tuesday or Wednesday if you are
willing to wait.

**Comment by Kunde21** [comment]
@vladimir-ch  That sounds good.  It's not adding functionality, so it's not a big rush.

**Comment by sbinet** [comment]
aside comment:

could we always rebase before merging?

with the mono-repo structure and with more people (yay!) working on different parts of the mono-repo, if we don't rebase before merging, we'll get a mess of a history.
it's already starting to be quite the spaghetti monster:

```
$> git l
*   5779c127 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #209 from gonum/f32/dot
|\  
| * 9d7baa81 asm/f32: Fixes from PR comments.
| * e0a0508f asm/f32: Asm dot functions.
| * 7f0e3341 asm/f32: Adding tests for dot asm functions.
| * 51fc8a3f asm/f32:  Adding stubs for dot asm functions.
| | * ab2c1e6e (origin/flow) Revert "graph/path: implement sophisticated algorithm"
| | * a9929ba0 graph/path: implement sophisticated algorithm
| | * 09fc4a20 graph/path: add benchmarks for Dominators
| | * bd6dc762 graph/path: replace map and slice index chasing with pointer chasing
| | * c5923049 Revert "graph/path: replace recursive dfs with iterative implementation"
| | * 7a3afb07 graph/path: add dominator tree API
| | * 4c195282 graph/path: replace recursive dfs with iterative implementation
| | * a4553c6e graph/path: replace Dominators with naive Lengauer and Tarjan
| | | * a6e52c49 (origin/kconnected) graph/community: add k-clique community function
| | | * a1136aba graph/topo: add clique graph construction function
| | |/  
| | | * 70474858 (origin/f64/ger) asm/f64: Merging logic changes from #225
| | | * 9dd91876 blas/gonum, asm/f64: Incorporate Ger routine in blas for testing.
| | | * e11ffa2c Build flags for go1.7 testing and guards for oversized test.
| | | * a9c13538 asm/f64: Cleaned up with go1.8-supported instructions.
| | | * 59c56216 asm/f64: Ger asm function added.
| | | * e82f927d asm/f64: Ger asm function framed out for loop construct/flow.
| | | * f8ea1eb8 asm/f64:  Adding stubs for Ger asm implementation.
| |_|/  
|/| |   
* | |   4ac7663e Merge pull request #227 from gonum/generate-error-fix
|\ \ \  
| * | | 1fc51d6a travis: Print git diff when check-generate fails.
* | | | 34774386 blas/gonum:  Fix NaN popagation in Dger.  (#225)
|/ / /  
* | | 49f3edb2 travis: reduce verbosity of test output
* | | f01292bb floats: Fix Argsort documentation.
* | | e01a71d4 mat: make RowView and ColView return Vector type and add RowViewOf and ColViewOf
| |/  
|/|   
| | * 134c8aaf (origin/negative-cycles) graph/path: allow Shortest to traverse negative cycles
| |/  
|/|   
* | d7342e68 travis: add go-1.9.x
* | 40dbcd03 travis: remove go-1.6.x
| | * 767fa7a0 (origin/degeneracy) graph/topo: use degeneracyOrdering instead of VertexOrdering for BronKerbosch
| | * 781be0e4 graph/topo: fix error reporting for VertexOrdering test
| | * 93093475 graph/topo: add function to find k-core of undirected graph
| | * 12013324 graph/{internal,path,topo}: factor reverse function out into internal
| |/  
|/|   
```

compare this to the Go stdlib mono-repo:

```
$> git l
* 0202aa8ec7 (origin/master, origin/HEAD) doc: update DWARF version
* 13772edbbd cmd/compile: simplify exporting universal 'error' type
* 261a8d9abd testing: use time.Since instead of time.Now().Sub
* 6367c19f26 cmd/internal/obj/x86: add some more AVX2 instructions
* 31ddd8a39d cmd/compile: more compact DWARF location for locals and arguments
* 84188296e9 cmd/compile: more compact representation of DW_AT_high_pc
* 50f1f639a4 cmd/asm: add most SSE4 missing instructions
* b15e8babc8 cmd/asm: add amd64 PALIGNR instruction
* 4074e4e5be cmd/asm: add amd64 CLFLUSH instruction
* 26dadbe32c cmd/asm: add amd64 PAB{SB,SD,SW}, PMADDUBSW, PMULHRSW, PSIG{NB,ND,NW}
* 16edf0b1f7 crypto/cipher: panic when IV length does not equal block size in NewOFB
* 6c102e141c cmd/compile: avoid stack allocation of a map bucket for large constant hints
* 034d825ea3 runtime: avoid redundant zeroing of hiter
* 9c3f268558 cmd/compile: set hiter type for map iterator in order pass
* bb31217579 runtime: move map ismapkey check to the compiler
* 2d362f7a49 flag: simplify arg logic in parseOne
* 6675fadfb8 runtime: cleanup amd64p32 memmove and memclr file organization
* 0fb82fbcce cmd/compile: remove global bout variable
* 34db5f0c4d cmd/compile: fix evaluation order for OASOP
* d349fa25df cmd/compile: fix and improve struct field reflect information
* 812b34efae cmd/compile: ignore non-code nodes when inlining
* b2e8630f2f cmd/compile: simplify "missing function body" error message
* 53d24f76fb doc: Fixed missing dot in effective_go.html
* aed1c119fd cmd/compile: fix assembly test
* fb165eaffd cmd/compile: combine x*n - y*n into (x-y)*n
* 03c3bb5f84 math: Add Round function (ties away from zero)
```

much more linear. much more readable.

**Comment by Kunde21** [comment]
A rebase after approval would wipe out the review, so we might end up chasing each other around trying to rebase before merging.  Squashing merges might be a better option to keep the history clean.

**Comment by sbinet** [comment]
(let's move this to `gonum-dev`)

**Pull request merged**